Error on CI using StatelessSession#fetch

When using Oracle, sometimes we have an exception because an object is a CompletionStage instead of the exepcted instance.

It seems to only happen with Oracle and only when some conditions are met (not sure exactly which).

This is the full stacktrace:
```
Gradle Test Executor 1 > org.hibernate.reactive.ReactiveStatelessProxyUpdateTest > testUpdateWithInitializedProxy STANDARD_OUT
    ERROR errors [vert.x-eventloop-thread-0] HR000057: Failed to execute statement [$1select reactivest0_.id as id1_0_0_, reactivest0_.sample_field as sample_field2_0_0_ from sample_entities reactivest0_ where reactivest0_.id=?]: $2could not load an entity: [org.hibernate.reactive.ReactiveStatelessProxyUpdateTest$SampleEntity#2]
    java.util.concurrent.CompletionException: org.hibernate.PropertyAccessException: Could not set field value [test] value by reflection : [class org.hibernate.reactive.ReactiveStatelessProxyUpdateTest$SampleEntity.sampleField] setter of org.hibernate.reactive.ReactiveStatelessProxyUpdateTest$SampleEntity.sampleField
    	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314) ~[?:?]
    	at java.util.concurrent.CompletableFuture.uniAcceptNow(CompletableFuture.java:757) ~[?:?]
    	at java.util.concurrent.CompletableFuture.uniAcceptStage(CompletableFuture.java:731) ~[?:?]
    	at java.util.concurrent.CompletableFuture.thenAccept(CompletableFuture.java:2108) ~[?:?]
    	at java.util.concurrent.CompletableFuture.thenAccept(CompletableFuture.java:143) ~[?:?]
    	at org.hibernate.reactive.loader.ReactiveResultSetProcessor.initializeEntity(ReactiveResultSetProcessor.java:80) ~[main/:?]
    	at org.hibernate.reactive.loader.entity.impl.ReactivePlanEntityLoader$ReactiveRowReader.lambda$reactivePerformTwoPhaseLoad$1(ReactivePlanEntityLoader.java:441) ~[main/:?]
    	at org.hibernate.reactive.util.impl.CompletionStages.lambda$loop$4(CompletionStages.java:324) ~[main/:?]
    	at org.hibernate.reactive.util.impl.CompletionStages$ArrayLoop.next(CompletionStages.java:396) ~[main/:?]
    	at org.hibernate.reactive.util.async.impl.AsyncTrampoline.lambda$asyncWhile$1(AsyncTrampoline.java:215) ~[main/:?]
    	at org.hibernate.reactive.util.async.impl.AsyncTrampoline$TrampolineInternal.unroll(AsyncTrampoline.java:121) ~[main/:?]
    	at org.hibernate.reactive.util.async.impl.AsyncTrampoline$TrampolineInternal.trampoline(AsyncTrampoline.java:102) ~[main/:?]
    	at org.hibernate.reactive.util.async.impl.AsyncTrampoline.asyncWhile(AsyncTrampoline.java:197) ~[main/:?]
    	at org.hibernate.reactive.util.async.impl.AsyncTrampoline.asyncWhile(AsyncTrampoline.java:215) ~[main/:?]
    	at org.hibernate.reactive.util.impl.CompletionStages.loop(CompletionStages.java:356) ~[main/:?]
    	at org.hibernate.reactive.util.impl.CompletionStages.loop(CompletionStages.java:324) ~[main/:?]
    	at org.hibernate.reactive.util.impl.CompletionStages.loop(CompletionStages.java:320) ~[main/:?]
    	at org.hibernate.reactive.loader.entity.impl.ReactivePlanEntityLoader$ReactiveRowReader.reactivePerformTwoPhaseLoad(ReactivePlanEntityLoader.java:439) ~[main/:?]
    	at org.hibernate.reactive.loader.entity.impl.ReactivePlanEntityLoader$ReactiveRowReader.reactiveFinishUp(ReactivePlanEntityLoader.java:408) ~[main/:?]
    	at org.hibernate.reactive.loader.entity.impl.ReactivePlanEntityLoader$ReactiveLoadPlanBasedResultSetProcessor.reactiveExtractResults(ReactivePlanEntityLoader.java:354) ~[main/:?]
    	at org.hibernate.reactive.loader.ReactiveLoader.reactiveProcessResultSet(ReactiveLoader.java:145) ~[main/:?]
    	at org.hibernate.reactive.loader.ReactiveLoader.lambda$doReactiveQueryAndInitializeNonLazyCollections$0(ReactiveLoader.java:77) ~[main/:?]
    	at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1072) ~[?:?]
    	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506) ~[?:?]
    	at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2073) ~[?:?]
    	at io.vertx.core.Future.lambda$toCompletionStage$2(Future.java:360) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.core.impl.future.FutureImpl$3.onSuccess(FutureImpl.java:141) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:60) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:211) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.core.impl.future.PromiseImpl.tryComplete(PromiseImpl.java:23) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.sqlclient.impl.QueryResultBuilder.tryComplete(QueryResultBuilder.java:102) ~[vertx-sql-client-4.2.7.jar:4.2.7]
    	at io.vertx.sqlclient.impl.QueryResultBuilder.tryComplete(QueryResultBuilder.java:35) ~[vertx-sql-client-4.2.7.jar:4.2.7]
    	at io.vertx.core.Promise.complete(Promise.java:66) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.core.Promise.handle(Promise.java:51) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.core.Promise.handle(Promise.java:29) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.core.impl.future.FutureImpl$3.onSuccess(FutureImpl.java:141) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.core.impl.future.FutureBase.lambda$emitSuccess$0(FutureBase.java:54) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.core.impl.EventLoopContext.execute(EventLoopContext.java:81) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.core.impl.future.PromiseImpl.tryComplete(PromiseImpl.java:23) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.core.impl.future.PromiseImpl.onSuccess(PromiseImpl.java:49) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.core.impl.future.PromiseImpl.handle(PromiseImpl.java:41) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.sqlclient.impl.TransactionImpl.lambda$wrap$0(TransactionImpl.java:72) ~[vertx-sql-client-4.2.7.jar:4.2.7]
    	at io.vertx.core.impl.future.FutureImpl$3.onSuccess(FutureImpl.java:141) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.core.impl.future.FutureBase.lambda$emitSuccess$0(FutureBase.java:54) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.core.impl.EventLoopContext.execute(EventLoopContext.java:81) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.core.impl.ContextImpl.execute(ContextImpl.java:260) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.core.impl.EventLoopContext.execute(EventLoopContext.java:22) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:51) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:211) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.core.impl.future.Transformation$1.onSuccess(Transformation.java:63) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.core.impl.future.FutureBase.lambda$emitSuccess$0(FutureBase.java:54) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.core.impl.EventLoopContext.execute(EventLoopContext.java:81) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.core.impl.ContextImpl.execute(ContextImpl.java:260) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.core.impl.EventLoopContext.execute(EventLoopContext.java:22) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:51) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.core.impl.future.FutureImpl.addListener(FutureImpl.java:196) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.core.impl.future.PromiseImpl.addListener(PromiseImpl.java:23) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.core.impl.future.Transformation.onSuccess(Transformation.java:44) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.core.impl.future.FutureImpl$ListenerArray.onSuccess(FutureImpl.java:262) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:60) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:211) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.core.impl.future.FixedMapping.onSuccess(FixedMapping.java:31) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.core.impl.future.FutureImpl$ListenerArray.onSuccess(FutureImpl.java:262) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:60) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:211) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.core.impl.future.Composition$1.onSuccess(Composition.java:62) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.core.impl.future.FutureImpl$ListenerArray.onSuccess(FutureImpl.java:262) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:60) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:211) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.core.impl.future.Composition$1.onSuccess(Composition.java:62) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:60) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.core.impl.future.SucceededFuture.addListener(SucceededFuture.java:88) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.core.impl.future.Composition.onSuccess(Composition.java:43) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:60) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:211) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.core.impl.future.PromiseImpl.tryComplete(PromiseImpl.java:23) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.oracleclient.impl.Helper$1.lambda$onNext$0(Helper.java:93) ~[vertx-oracle-client-4.2.7.jar:4.2.7]
    	at io.vertx.core.impl.AbstractContext.dispatch(AbstractContext.java:100) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.core.impl.AbstractContext.dispatch(AbstractContext.java:63) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.vertx.core.impl.EventLoopContext.lambda$runOnContext$0(EventLoopContext.java:38) ~[vertx-core-4.2.7.jar:4.2.7]
    	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164) [netty-common-4.1.74.Final.jar:4.1.74.Final]
    	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:469) [netty-common-4.1.74.Final.jar:4.1.74.Final]
    	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:503) [netty-transport-4.1.74.Final.jar:4.1.74.Final]
    	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986) [netty-common-4.1.74.Final.jar:4.1.74.Final]
    	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) [netty-common-4.1.74.Final.jar:4.1.74.Final]
    	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) [netty-common-4.1.74.Final.jar:4.1.74.Final]
    	at java.lang.Thread.run(Thread.java:829) [?:?]
    Caused by: org.hibernate.WrongClassException: Object [id=2] was not of the specified subclass [org.hibernate.reactive.ReactiveStatelessProxyUpdateTest$SampleEntity] : loaded object was of wrong class class java.util.concurrent.CompletableFuture
    	at org.hibernate.loader.plan.exec.process.internal.EntityReferenceInitializerImpl.hydrateEntityState(EntityReferenceInitializerImpl.java:195) ~[hibernate-core-5.6.9.Final.jar:5.6.9.Final]
    	at org.hibernate.loader.plan.exec.process.internal.AbstractRowReader.readRow(AbstractRowReader.java:80) ~[hibernate-core-5.6.9.Final.jar:5.6.9.Final]
    	at org.hibernate.loader.plan.exec.process.internal.ResultSetProcessorImpl.extractRows(ResultSetProcessorImpl.java:157) ~[hibernate-core-5.6.9.Final.jar:5.6.9.Final]
    	at org.hibernate.reactive.loader.entity.impl.ReactivePlanEntityLoader$ReactiveLoadPlanBasedResultSetProcessor.reactiveExtractResults(ReactivePlanEntityLoader.java:352) ~[main/:?]
    	at org.hibernate.reactive.loader.ReactiveLoader.reactiveProcessResultSet(ReactiveLoader.java:145) ~[main/:?]
    	at org.hibernate.reactive.loader.ReactiveLoader.lambda$doReactiveQueryAndInitializeNonLazyCollections$0(ReactiveLoader.java:77) ~[main/:?]
    	at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1072) ~[?:?]
    	... 68 more
```

--- a/hibernate-reactive-core/src/main/java/org/hibernate/reactive/logging/impl/Log.java
+++ b/hibernate-reactive-core/src/main/java/org/hibernate/reactive/logging/impl/Log.java
@@ -174,7 +174,7 @@ public interface Log extends BasicLogger {
 	LazyInitializationException collectionCannotBeInitializedlazyInitializationException(String role);
 
 	@LogMessage(level = ERROR)
-	@Message(id = 57, value = "Failed to execute statement [$1%s]: $2%s")
+	@Message(id = 57, value = "Failed to execute statement [%1$s]: %2$s")
 	void failedToExecuteStatement(String sql, String s, @Cause Throwable t);
 
 	@Message(id = 58, value = "Impossible flush mode")

--- a/hibernate-reactive-core/src/main/java/org/hibernate/reactive/session/impl/ReactiveStatelessSessionImpl.java
+++ b/hibernate-reactive-core/src/main/java/org/hibernate/reactive/session/impl/ReactiveStatelessSessionImpl.java
@@ -746,23 +746,33 @@ public class ReactiveStatelessSessionImpl extends StatelessSessionImpl
 			else {
 				String entityName = initializer.getEntityName();
 				Serializable id = initializer.getIdentifier();
-				ReactiveEntityPersister persister = (ReactiveEntityPersister)
-						getFactory().getMetamodel().entityPersister( entityName );
+				ReactiveEntityPersister persister = (ReactiveEntityPersister) getFactory()
+						.getMetamodel().entityPersister( entityName );
 				initializer.setSession( this );
 				persistenceContext.beforeLoad();
-				return persister.reactiveLoad( id, initializer.getImplementation(), LockOptions.NONE, this )
-						.whenComplete( (v, e) -> {
-							persistenceContext.afterLoad();
-							if ( persistenceContext.isLoadFinished() ) {
-								persistenceContext.clear();
-							}
-						} )
-						.thenApply( entity -> {
-							checkEntityFound( this, entityName, id, entity );
-							initializer.setImplementation( entity );
-							initializer.unsetSession();
-							return unproxy ? (T) entity : association;
-						} );
+
+				// This is hard to test because it happens on slower machines like the ones we use on CI.
+				// See AbstractLazyInitializer#initialize, it happens when the object is not initialized and we need to
+				// call session.immediateLoad
+				CompletionStage<?> stage = initializer.getImplementation() instanceof CompletionStage
+					? (CompletionStage<?>) initializer.getImplementation()
+					: completedFuture( initializer.getImplementation() );
+
+				return stage
+						.thenCompose( implementation -> persister
+								.reactiveLoad( id, implementation, LockOptions.NONE, this )
+								.whenComplete( (v, e) -> {
+									persistenceContext.afterLoad();
+									if ( persistenceContext.isLoadFinished() ) {
+										persistenceContext.clear();
+									}
+								} )
+								.thenApply( entity -> {
+									checkEntityFound( this, entityName, id, entity );
+									initializer.setImplementation( entity );
+									initializer.unsetSession();
+									return unproxy ? (T) entity : association;
+								} ) );
 			}
 		}
 		else if ( association instanceof PersistentCollection ) {

--- a/hibernate-reactive-core/src/test/java/org/hibernate/reactive/ReactiveStatelessProxyUpdateTest.java
+++ b/hibernate-reactive-core/src/test/java/org/hibernate/reactive/ReactiveStatelessProxyUpdateTest.java
@@ -5,9 +5,9 @@
  */
 package org.hibernate.reactive;
 
-import java.io.Serializable;
 import java.util.Collection;
 import java.util.List;
+import java.util.function.IntFunction;
 import javax.persistence.Column;
 import javax.persistence.Entity;
 import javax.persistence.FetchType;
@@ -20,9 +20,11 @@ import javax.persistence.Table;
 
 import org.hibernate.HibernateException;
 import org.hibernate.LazyInitializationException;
+import org.hibernate.reactive.util.impl.CompletionStages;
 
 import org.junit.Test;
 
+import io.smallrye.mutiny.Uni;
 import io.vertx.ext.unit.TestContext;
 
 import static org.assertj.core.api.Assertions.assertThat;
@@ -31,8 +33,8 @@ import static org.hibernate.reactive.testing.ReactiveAssertions.assertThrown;
 /**
  * Test the stateless update of a proxy.
  * <p>
- *     Note that it's required to update and read the values from the proxy using getter/setter and
- *     there is no guarantee about this working otherwise.
+ * Note that it's required to update and read the values from the proxy using getter/setter and
+ * there is no guarantee about this working otherwise.
  * </p>
  *
  * @see org.hibernate.reactive.session.impl.ReactiveStatelessSessionImpl#reactiveUpdate(Object)
@@ -40,32 +42,35 @@ import static org.hibernate.reactive.testing.ReactiveAssertions.assertThrown;
  */
 public class ReactiveStatelessProxyUpdateTest extends BaseReactiveTest {
 
+	/**
+	 * Number of updates. On slower machines like CI this test will fail even with a slow number of entries.
+	 * On faster machines, it needs a lot more updates.
+	 */
+	private final static int UPDATES_NUM = 10;
+
 	@Override
 	protected Collection<Class<?>> annotatedEntities() {
-		return List.of( SampleEntity.class, SampleJoinEntity.class );
+		return List.of( Game.class, GameCharacter.class );
 	}
 
 	@Test
 	public void testUnfetchedEntityException(TestContext context) {
-		SampleEntity sampleEntity = new SampleEntity();
-		sampleEntity.sampleField = "test";
-
-		SampleJoinEntity sampleJoinEntity = new SampleJoinEntity();
-		sampleJoinEntity.sampleEntity = sampleEntity;
+		Game lol = new Game( "League of Legends" );
+		GameCharacter ck = new GameCharacter( "Caitlyn Kiramman" );
+		ck.setGame( lol );
 
 		test( context, assertThrown( HibernateException.class, getMutinySessionFactory()
-				.withTransaction( (s, t) -> s.persistAll( sampleEntity, sampleJoinEntity ) )
-				.chain( targetId -> getMutinySessionFactory()
-						.withTransaction( (session, transaction) -> session
-								.find( SampleJoinEntity.class, sampleJoinEntity.getId() ) )
+				.withTransaction( s -> s.persistAll( lol, ck ) )
+				.chain( targetId -> getMutinySessionFactory().withTransaction( session -> session
+						.find( GameCharacter.class, ck.getId() ) )
 				)
-				.chain( joinEntityFromDatabase -> getMutinySessionFactory().withStatelessTransaction(
-						(s, t) -> {
-							SampleEntity entityFromDb = joinEntityFromDatabase.sampleEntity;
-							entityFromDb.sampleField = "updated field";
+				.call( charFound -> getMutinySessionFactory()
+						.withStatelessTransaction( s -> {
+							Game game = charFound.getGame();
+							game.gameTitle = "League of Legends V2";
 							// We expect the update to fail because we haven't fetched the entity
 							// and therefore the proxy is uninitialized
-							return s.update( entityFromDb );
+							return s.update( game );
 						} )
 				) )
 				.invoke( exception -> assertThat( exception.getMessage() ).contains( "HR000072" ) )
@@ -74,73 +79,77 @@ public class ReactiveStatelessProxyUpdateTest extends BaseReactiveTest {
 
 	@Test
 	public void testLazyInitializationException(TestContext context) {
-		SampleEntity sampleEntity = new SampleEntity();
-		sampleEntity.sampleField = "test";
-
-		SampleJoinEntity sampleJoinEntity = new SampleJoinEntity();
-		sampleJoinEntity.sampleEntity = sampleEntity;
+		Game lol = new Game( "League of Legends" );
+		GameCharacter ck = new GameCharacter( "Caitlyn Kiramman" );
+		ck.setGame( lol );
 
 		test( context, assertThrown( LazyInitializationException.class, getMutinySessionFactory()
-				.withTransaction( (s, t) -> s.persistAll( sampleEntity, sampleJoinEntity ) )
-
+				.withTransaction( s -> s.persistAll( lol, ck ) )
 				.chain( targetId -> getMutinySessionFactory()
-						.withTransaction( (session, transaction) -> session
-								.find( SampleJoinEntity.class, sampleJoinEntity.getId() ) )
+						.withStatelessSession( session -> session.get( GameCharacter.class, ck.getId() ) )
 				)
-
-				.chain( joinEntityFromDatabase -> getMutinySessionFactory().withStatelessTransaction(
-						(s, t) -> {
+				.call( charFound -> getMutinySessionFactory()
+						.withStatelessTransaction( s -> {
+							Game game = charFound.getGame();
 							// LazyInitializationException here because we haven't fetched the entity
-							SampleEntity entityFromDb = joinEntityFromDatabase.getSampleEntity();
-							entityFromDb.setSampleField( "updated field" );
-							return s.update( entityFromDb );
+							game.setGameTitle( "League of Legends V2" );
+							context.fail( "We were expecting a LazyInitializationException" );
+							return null;
 						} )
-				) )
-		);
+				)
+		) );
 	}
 
 	@Test
-	public void testUpdateWithInitializedProxy(TestContext context) {
-		SampleEntity sampleEntity = new SampleEntity();
-		sampleEntity.setSampleField( "test" );
+	public void testUpdateWithInitializedProxyInLoop(TestContext context) {
+		test( context, loop( 0, UPDATES_NUM, i -> {
+			Game lol = new Game( "League of Legends" );
+			GameCharacter ck = new GameCharacter( "Caitlyn Kiramman" );
+			ck.setGame( lol );
 
-		SampleJoinEntity sampleJoinEntity = new SampleJoinEntity();
-		sampleJoinEntity.setSampleEntity( sampleEntity );
-
-		test( context, getMutinySessionFactory()
-					  .withTransaction( (s, t) -> s.persistAll( sampleEntity, sampleJoinEntity ) )
-
-					  .chain( targetId -> getMutinySessionFactory()
-							  .withTransaction( (session, transaction) -> session
-									  .find( SampleJoinEntity.class, sampleJoinEntity.getId() ) )
-					  )
-
-					  .chain( joinEntityFromDatabase -> getMutinySessionFactory().withStatelessTransaction(
-							  (s, t) -> s
-									  // The update of the associated entity should work if we fetch it first
-									  .fetch( joinEntityFromDatabase.getSampleEntity() )
-									  .chain( fetchedEntity -> {
-										  fetchedEntity.setSampleField( "updated field" );
-										  return s.update( fetchedEntity );
-									  } ) )
-					  )
-					  .chain( () -> getMutinySessionFactory().withSession( session -> session
-							  .createQuery( "from SampleEntity", SampleEntity.class )
-							  .getSingleResult()
-							  .onItem().invoke( result -> context.assertEquals( "updated field", result.getSampleField() ) ) )
-					  )
-		);
+			return getMutinySessionFactory()
+					.withTransaction( s -> s.persistAll( lol, ck ) )
+					.chain( targetId -> getMutinySessionFactory()
+							.withSession( session -> session.find( GameCharacter.class, ck.getId() ) )
+					)
+					.chain( charFound -> getMutinySessionFactory()
+							.withStatelessTransaction( s -> s
+									// The update of the associated entity should work if we fetch it first
+									.fetch( charFound.getGame() )
+									.chain( fetchedGame -> {
+										fetchedGame.setGameTitle( "League of Legends V2" );
+										return s.update( fetchedGame );
+									} )
+							)
+					)
+					.call( () -> getMutinySessionFactory().withSession( session -> session
+							.find( Game.class, lol.getId() )
+							.invoke( result -> context.assertEquals( "League of Legends V2", result.getGameTitle() ) ) )
+					);
+		} ) );
 	}
 
-	@Entity(name = "SampleEntity")
-	@Table(name = "sample_entities")
-	public static class SampleEntity implements Serializable {
+	private Uni<Void> loop(int start, int end, IntFunction<Uni<Void>> intFun) {
+		return Uni.createFrom()
+				.completionStage( CompletionStages.loop( start, end, index -> intFun.apply( index ).subscribeAsCompletionStage() ) );
+	}
+
+	@Entity(name = "Game")
+	@Table(name = "game")
+	public static class Game {
 		@Id
 		@GeneratedValue(strategy = GenerationType.IDENTITY)
 		private Long id;
 
-		@Column(name = "sample_field")
-		private String sampleField;
+		@Column(name = "title")
+		private String gameTitle;
+
+		public Game() {
+		}
+
+		public Game(String gameTitle) {
+			this.gameTitle = gameTitle;
+		}
 
 		public Long getId() {
 			return id;
@@ -150,30 +159,39 @@ public class ReactiveStatelessProxyUpdateTest extends BaseReactiveTest {
 			this.id = id;
 		}
 
-		public String getSampleField() {
-			return sampleField;
+		public String getGameTitle() {
+			return gameTitle;
 		}
 
-		public void setSampleField(String sampleField) {
-			this.sampleField = sampleField;
+		public void setGameTitle(String gameTitle) {
+			this.gameTitle = gameTitle;
 		}
 
 		@Override
 		public String toString() {
-			return getClass().getSimpleName() + " ID: " + id + " sampleField: " + sampleField;
+			return id + ":" + gameTitle;
 		}
 	}
 
-	@Entity(name = "SampleJoinEntity")
-	@Table(name = "sample_join_entities")
-	public static class SampleJoinEntity implements Serializable {
+	@Entity(name = "GameCharacter")
+	@Table(name = "game_character")
+	public static class GameCharacter {
 		@Id
 		@GeneratedValue(strategy = GenerationType.IDENTITY)
 		private Long id;
+
+		private String name;
 
 		@ManyToOne(fetch = FetchType.LAZY)
-		@JoinColumn(name = "sample_entity_id", referencedColumnName = "id")
-		private SampleEntity sampleEntity;
+		@JoinColumn(name = "game_id", referencedColumnName = "id")
+		private Game game;
+
+		public GameCharacter() {
+		}
+
+		public GameCharacter(String name) {
+			this.name = name;
+		}
 
 		public Long getId() {
 			return id;
@@ -183,17 +201,25 @@ public class ReactiveStatelessProxyUpdateTest extends BaseReactiveTest {
 			this.id = id;
 		}
 
-		public SampleEntity getSampleEntity() {
-			return sampleEntity;
+		public String getName() {
+			return name;
 		}
 
-		public void setSampleEntity(SampleEntity sampleEntity) {
-			this.sampleEntity = sampleEntity;
+		public void setName(String name) {
+			this.name = name;
+		}
+
+		public Game getGame() {
+			return game;
+		}
+
+		public void setGame(Game game) {
+			this.game = game;
 		}
 
 		@Override
 		public String toString() {
-			return getClass().getSimpleName() + " ID: " + id + " sampleEntity: " + sampleEntity;
+			return id + ":" + name + ":" + game;
 		}
 	}
 }

